File: lib/domain/weather_data/interactors/fetch_weather_data_retry_on_interactor.rb
# frozen_string_literal: true

module Domain
  module WeatherData
    module Interactors
      class FetchWeatherDataRetryOnInteractor
        include InputPorts::FetchWeatherDataRetryOnInputPort

        def initialize(farm_gateway:, presenter:, advance_phase_interactor:, mark_farm_weather_data_failed_interactor:, logger:, translator:)
          @farm_gateway = farm_gateway
          @presenter = presenter
          @advance_phase_interactor = advance_phase_interactor
          @mark_farm_weather_data_failed_interactor = mark_farm_weather_data_failed_interactor
          @logger = logger
          @translator = translator
        end

        def call(input_dto:)
          farm_id = input_dto[:farm_id]
          start_date = input_dto[:start_date]
          end_date = input_dto[:end_date]
          executions = input_dto[:executions]
          error_message = input_dto[:error_message]
          period_str = start_date.year == end_date.year ? "#{start_date.year}" : "#{start_date.year}-#{end_date.year}"

          @presenter.error "❌ [Farm##{farm_id}] Failed to fetch weather data for #{period_str} after #{executions} attempts"
          @presenter.error "   Final error: #{error_message}"

          error_msg = @translator.t("jobs.fetch_weather_data.retry_limit_exceeded", error: error_message)
          if farm_id
            @mark_farm_weather_data_failed_interactor.call(
              Domain::Farm::Dtos::MarkFarmWeatherDataFailedInput.new(
                farm_id: farm_id,
                error_message: error_msg
              )
            )
          end

          if input_dto[:cultivation_plan_id] && input_dto[:channel_class]
            @advance_phase_interactor.call(
              Domain::CultivationPlan::Dtos::AdvanceCultivationPlanPhaseInput.new(
                plan_id: input_dto[:cultivation_plan_id],
                phase_name: :phase_failed,
                channel_class: input_dto[:channel_class],
                failure_subphase: "fetching_weather"
              )
            )
          end
        end
      end
    end
  end
end

File: crates/agrr-domain/src/weather_data/interactors/fetch_weather_data_retry_on_interactor.rs
//! Ruby: `Domain::WeatherData::Interactors::FetchWeatherDataRetryOnInteractor`

use std::collections::BTreeMap;

use crate::shared::ports::{LoggerPort, TranslatorPort};
use crate::weather_data::dtos::FetchWeatherDataRetryOnInput;
use crate::weather_data::ports::{
    AdvanceCultivationPlanPhasePort, FetchWeatherDataJobPresenterPort,
    MarkFarmWeatherDataFailedPort,
};

/// Ruby: `Domain::WeatherData::Interactors::FetchWeatherDataRetryOnInteractor`
pub struct FetchWeatherDataRetryOnInteractor<'a, T> {
    presenter: &'a dyn FetchWeatherDataJobPresenterPort,
    advance_phase: &'a dyn AdvanceCultivationPlanPhasePort,
    mark_failed: &'a dyn MarkFarmWeatherDataFailedPort,
    logger: &'a dyn LoggerPort,
    translator: &'a T,
}

impl<'a, T> FetchWeatherDataRetryOnInteractor<'a, T>
where
    T: TranslatorPort,
{
    pub fn new(
        presenter: &'a dyn FetchWeatherDataJobPresenterPort,
        advance_phase: &'a dyn AdvanceCultivationPlanPhasePort,
        mark_failed: &'a dyn MarkFarmWeatherDataFailedPort,
        logger: &'a dyn LoggerPort,
        translator: &'a T,
    ) -> Self {
        Self {
            presenter,
            advance_phase,
            mark_failed,
            logger,
            translator,
        }
    }

    pub fn call(&self, input: FetchWeatherDataRetryOnInput) {
        let farm_id = input.farm_id;
        let period_str = period_label(input.start_date, input.end_date);

        self.presenter.error(&format!(
            "❌ [Farm#{:?}] Failed to fetch weather data for {period_str} after {} attempts",
            farm_id, input.executions
        ));
        self.presenter
            .error(&format!("   Final error: {}", input.error_message));

        let mut options = BTreeMap::new();
        options.insert("error".to_string(), input.error_message.clone());
        let error_msg = self
            .translator
            .t("jobs.fetch_weather_data.retry_limit_exceeded", &options);

        if let Some(farm_id) = farm_id {
            self.mark_failed.call(farm_id, &error_msg);
        }

        if let (Some(plan_id), Some(channel)) =
            (input.cultivation_plan_id, input.channel_class.as_deref())
        {
            self.advance_phase
                .call_failed_fetching_weather(plan_id, channel);
        }

        let _ = self.logger;
    }
}

fn period_label(start_date: time::Date, end_date: time::Date) -> String {
    if start_date.year() == end_date.year() {
        start_date.year().to_string()
    } else {
        format!("{}-{}", start_date.year(), end_date.year())
    }
}

#[cfg(test)]
mod interactors_fetch_weather_data_retry_on_interactor_test_inline {
    use super::*;
    include!(concat!(env!("CARGO_MANIFEST_DIR"), "/test/weather_data/interactors_fetch_weather_data_retry_on_interactor_test.rs"));
}
